### PR TITLE
Review category

### LINKED ISSUE #4776: Peer review: Do not ask to review categories if there are none
If a picture has no category, then we already know the answer: The picture is not well-categorized.
In such cases, we could skip that step.
Even if some bots routinely flag all non-categorized pictures, our mobile peer-review still makes sense for other aspects of the pictures (encyclopedic or not, relevant categories or not, etc).

### PR DESCRIPTION
**Pull request by devarsh-mavani-19**
**Description (required)**
In peer review, If a picture has no category, then we already know the answer: The picture is not well-categorized.
In such cases, we could skip that step.

Fixes #4776 

What changes did you make and why?
I checked if category exist then show that screen else skip that step
**Tests performed (required)**
Build Variant: **ProdDebug**
Device: Redmi 6 Pro

### CODE REVIEW

**Comment by devarsh-mavani-19** [comment]
hey, @nicolas-raoul the issue is fixed. can you check this, please? Thanks

File: app/src/main/java/fr/free/nrw/commons/review/ReviewActivity.java [hunk 1]
```diff
@@ -71,6 +81,8 @@
      */
      private ReviewImageFragment reviewImageFragment;
 
+     boolean categoryExists = false;
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/review/ReviewActivity.java:84]: Could you please rename to `hasNonHiddenCategories`?

Comment (nicolas-raoul) [reply app/src/main/java/fr/free/nrw/commons/review/ReviewActivity.java:84]: Also, please make it private and add javadoc.

File: app/src/main/java/fr/free/nrw/commons/review/ReviewActivity.java [hunk 2]
```diff
@@ -153,15 +165,33 @@ public boolean onSupportNavigateUp() {
...
         compositeDisposable.add(reviewHelper.getRandomMedia()
                 .subscribeOn(Schedulers.io())
                 .observeOn(AndroidSchedulers.mainThread())
                 .subscribe(media -> {
-                    reviewImageFragment = getInstanceOfReviewImageFragment();
-                    reviewImageFragment.disableButtons();
-                    updateImage(media);
+                    Service service = ServiceFactory.get(new WikiSite(Service.COMMONS_URL)
+                        , Service.COMMONS_URL, Service.class);
+
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/review/ReviewActivity.java:177]: Please add a `//`-type comment explaining what this block does.

File: app/src/main/java/fr/free/nrw/commons/review/ReviewActivity.java [hunk 3]
```diff
@@ -195,8 +225,15 @@ private void updateImage(Media media) {
 
     public void swipeToNext() {
         int nextPos = reviewPager.getCurrentItem() + 1;
+        // If currently at category fragment, then check if category exists
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/review/ReviewActivity.java:228]: `check if category exists` could be ambiguous for a newcomer.
I suggest `check whether the media has any non-hidden category`

File: app/src/main/java/fr/free/nrw/commons/review/ReviewActivity.java [hunk 4]
```diff
@@ -195,8 +225,15 @@ private void updateImage(Media media) {
...
         if (nextPos <= 3) {
             reviewPager.setCurrentItem(nextPos);
+            if (nextPos == 2) {
+                if (!categoryExists) {
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/review/ReviewActivity.java:232]: A comment could be developer-friendly here:
```
// The media has no non-hidden category. Such media are already flagged by server-side bots, so no need to review manually.
```

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
Minor requests.
Thanks! :-)

**Comment by devarsh-mavani-19** [comment]
@nicolas-raoul did the changes. Thanks : )

**Review by nicolas-raoul — COMMENTED** [review]
I just tested.
Shouldn't this `Oh, this is not even categorized!` feature be removed, and the category review step be skipped altogether?
adb

**Comment by devarsh-mavani-19** [comment]
> I just tested. Shouldn't this `Oh, this is not even categorized!` feature be removed, and the category review step be skipped altogether? adb

@nicolas-raoul I just checked the file that you have attached in screenshot and it actually has a non-hidden category. here is the link to it. 
[https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&clprop=hidden&prop=categories&titles=File:Stamp of Ajman - 1968 - Colnect 682825 - Ferdinand of Austria - by Velasquez and - Return from hunting.jpeg](https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&clprop=hidden&prop=categories&titles=File:Stamp of Ajman - 1968 - Colnect 682825 - Ferdinand of Austria - by Velasquez and - Return from hunting.jpeg
)

So I think that's why the screen was not skipped because there are actually some categories that are non-hidden. And I think "Oh this is not even categorized" is probably a separate bug. I will look into it and let you know if i find anything.
Thanks : )

**Comment by devarsh-mavani-19** [comment]
@nicolas-raoul I further tested this and I found that the backend API is inconsistent, 
here is what is happening
1. API requests a file for review (Let's take **Pierwsze zarysy zoologii 1887 (93616195).jpg** in our example)
Ex: [This link is being requested](https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&prop=imageinfo&iiprop=url|extmetadata|user&&iiurlwidth=640&iiextmetadatafilter=DateTime|Categories|GPSLatitude|GPSLongitude|ImageDescription|DateTimeOriginal|Artist|LicenseShortName|LicenseUrl&titles=File:Pierwsze%20zarysy%20zoologii%201887%20(93616195).jpg)
The above URL returns an empty category list 
2. I am requesting a category list with hidden attribute info 
Ex: [See this url](https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&clprop=hidden&prop=categories&titles=File:Pierwsze%20zarysy%20zoologii%201887%20(93616195).jpg)
This URL responds with a bunch of categories for the same file as in step 1.

so both API calls are giving different results for the same resource. So further, I opened the Wikimedia page for the same file
check this page and the result is second Endpoint (the one I implemented in this PR).
https://commons.wikimedia.org/wiki/File:Pierwsze_zarysy_zoologii_1887_(93616195).jpg

**Comment by devarsh-mavani-19** [comment]
> Thanks @devarsh-mavani-19 for the investigation! When I open https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&prop=imageinfo&iiprop=url%7Cextmetadata%7Cuser&&iiurlwidth=640&iiextmetadatafilter=DateTime%7CCategories%7CGPSLatitude%7CGPSLongitude%7CImageDescription%7CDateTimeOriginal%7CArtist%7CLicenseShortName%7CLicenseUrl&titles=File:Pierwsze%20zarysy%20zoologii%201887%20(93616195).jpg I see `Categories` `value` `"CC-PD-Mark|Files from Polona|Images contributed by the National Library in Warsaw|PD-old-100-expired|Patrimonium (project)|Pierwsze zarysy zoologii (1887)"` isn't it the same categories as shown by https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&clprop=hidden&prop=categories&titles=File:Pierwsze%20zarysy%20zoologii%201887%20(93616195).jpg (just shown in a different way)? Or was it giving a different result when you first opened it, maybe because of some delay or slow bot work?

I checked that before I sent to you and it was showing empty in Categories attribute in response 😅. I tested again and I found once such case again.

Here is the link
https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&prop=imageinfo&iiprop=url%7Cextmetadata%7Cuser&&iiurlwidth=640&iiextmetadatafilter=DateTime%7CCategories%7CGPSLatitude%7CGPSLongitude%7CImageDescription%7CDateTimeOriginal%7CArtist%7CLicenseShortName%7CLicenseUrl&titles=File:Vincenzo%20de%27%20rossi,%20morte%20di%20laocoonte%20coi%20figli,%201570%20ca.%20(coll.%20priv.%20galleria%20del%20laocoonte,%20roma-londra)%2001.jpg

and this is the categories version
https://commons.wikimedia.org/w/api.php?format=json&formatversion=2&errorformat=plaintext&action=query&prop=categories&clprop=hidden&cllimit=500&titles=File:Vincenzo%20de%27%20rossi,%20morte%20di%20laocoonte%20coi%20figli,%201570%20ca.%20(coll.%20priv.%20galleria%20del%20laocoonte,%20roma-londra)%2001.jpg

both of them have different outputs.
image

image

**Comment by nicolas-raoul** [comment]
Now both are showing the categories, so there seems to bee some propagation delay.

Would it be possible to use the second URL instead? So that we know what are the correct categories. Thanks a lot!

**Comment by devarsh-mavani-19** [comment]
@nicolas-raoul I have done the necessary changes and issues is solved. Can you please check it? Thanks
By the way on inspecting the code further, I found that both the API are working as the should be and the reason the first URL was sometime giving us wrong category is because those categories are Images' Meta-data and not the one entered by user. But as you said I think the bots are eventually adding the user entered categories in the image meta-data that's why when you were checking those links it had perfect category. But now it won't be an issue. I have handled that from frontend by storing the user entered category in separate field.

File: app/src/main/java/fr/free/nrw/commons/Media.kt [hunk 5]
```diff
@@ -77,7 +78,11 @@ class Media constructor(
     var coordinates: LatLng? = null,
     var captions: Map<String, String> = emptyMap(),
     var descriptions: Map<String, String> = emptyMap(),
-    var depictionIds: List<String> = emptyList()
+    var depictionIds: List<String> = emptyList(),
+    /**
+     * Stores the mapping of category title to hidden attribute
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/Media.kt:83]: Could you please add an example key and example value?

File: app/src/main/java/fr/free/nrw/commons/Media.kt [hunk 6]
```diff
@@ -77,7 +78,11 @@ class Media constructor(
...
+     */
+    var detailedCategories: Map<String, String> = emptyMap() // This field was added to find non-hidden categories
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/Media.kt:85]: You can move this `//` comment's text to inside the `/**` comment.

File: app/src/main/java/fr/free/nrw/commons/review/ReviewImageFragment.java [hunk 7]
```diff
@@ -52,8 +54,20 @@ public void update(int position) {
...
         Media media = getReviewActivity().getMedia();
-        if (media != null && media.getCategories() != null && isAdded()) {
-            String catString = TextUtils.join(", ", media.getCategories());
+        if (media != null && media.getDetailedCategories() != null && isAdded()) {
+            // Filter category name attribute from all categories
+            List<String> categories = new ArrayList<>();
+            for(String key : media.getDetailedCategories().keySet()) {
+                String value = String.valueOf(key);
+                // Each category returned has a format like "Category:<some-category-name>"
+                // so remove the prefix "Category:"
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/review/ReviewImageFragment.java:63]: Nice!

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]

**Comment by devarsh-mavani-19** [comment]
@nicolas-raoul Did the changes? can you please check it? Thanks

File: app/src/main/java/fr/free/nrw/commons/Media.kt [hunk 8]
```diff
@@ -77,7 +78,13 @@ class Media constructor(
...
     var captions: Map<String, String> = emptyMap(),
     var descriptions: Map<String, String> = emptyMap(),
-    var depictionIds: List<String> = emptyList()
+    var depictionIds: List<String> = emptyList(),
+    /**
+     * This field was added to find non-hidden categories
+     * Stores the mapping of category title to hidden attribute
+     * Example: Mountains => false, CC-BY-SA-2.0 => true
+     */
+    var detailedCategories: Map<String, String> = emptyMap()
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/Media.kt:87]: How about making it `Map<String, Boolean>`?
And also (for readibility) rename to `categoriesHiddenStatus`
Finally, please add quotes around String examples to make it clear what is a String example and what is not: `"Mountains" => false, "CC-BY-SA-2.0" => true`

Thanks a lot! :-)

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]

File: app/src/main/java/fr/free/nrw/commons/Media.kt (hunk 8 above)
Comment (devarsh-mavani-19) [reply app/src/main/java/fr/free/nrw/commons/Media.kt:87]: @nicolas-raoul I tried adding the Map<String, Boolean> and also added a typeconverter for it but still I was getting 
"error: Cannot figure out how to save this field into database. You can consider adding a type converter for it." build error. That's why I kept it string because there is already a type converter for map<string, string>. Not sure why getting that error

Comment (nicolas-raoul) [reply app/src/main/java/fr/free/nrw/commons/Media.kt:87]: Would you mind creating a type converter? I believe it can be done in a few lines. Thanks a lot!

Comment (devarsh-mavani-19) [reply app/src/main/java/fr/free/nrw/commons/Media.kt:87]: @nicolas-raoul as I mentioned above, i already created type converter method in converter.java class still getting this error.

Comment (devarsh-mavani-19) [reply app/src/main/java/fr/free/nrw/commons/Media.kt:87]: > Would you mind creating a type converter? I believe it can be done in a few lines. Thanks a lot!

In the converters.java file I added this new method
```
@TypeConverter
    public static String mapObjectToString2(Map<String,Boolean> objectList) {
        return writeObjectToString(objectList);
    }
```
still getting same error

**Comment by devarsh-mavani-19** [comment]
@nicolas-raoul did those changes as well. Can you check it, please? Thanks : )

File: app/src/main/java/fr/free/nrw/commons/db/Converters.java [hunk 9]
```diff
@@ -79,11 +79,21 @@ public static String mapObjectToString(Map<String,String> objectList) {
         return writeObjectToString(objectList);
     }
 
+    @TypeConverter
+    public static String mapObjectToString2(Map<String,Boolean> objectList) {
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/db/Converters.java:83]: Is this still necessary?

Comment (nicolas-raoul) [reply app/src/main/java/fr/free/nrw/commons/db/Converters.java:83]: Ah yes I guess sorry!

**Comment by devarsh-mavani-19** [comment]
@madhurgupta10 I have a question. If i want to mock a private method then do I have to use powermock? here is what I am trying to achieve
I am running a test on function OnUpdateCategoriesQuestion
inside it there is a call to private method **getReviewActivity()**. Now I want to do something like this 
```
doReturn(mock(ReviewActivity::class.java).`when`(fragment).getReviewActivity()
```

but all resource on internet are referring to use powermock. can you please guide me? Thanks

**Comment by madhurgupta10** [comment]
> OnUpdateCategoriesQuestion

`when-then` only works for mocks. Here, fragment isn't a mock. 

Your `getReviewActivity()` method simply returns the activity or fragment's parent so I am not sure why you need to use mock here?

```
activity = Robolectric.buildActivity(ReviewActivity::class.java).create().get()
```

```
private ReviewActivity getReviewActivity() {
    return (ReviewActivity) requireActivity();
}
```

**Comment by devarsh-mavani-19** [comment]
> > OnUpdateCategoriesQuestion
> 
> `when-then` only works for mocks. Here, fragment isn't a mock.
> 
> Your `getReviewActivity()` method simply returns the activity or fragment's parent so I am not sure why you need to use mock here?
> 
> ```
> activity = Robolectric.buildActivity(ReviewActivity::class.java).create().get()
> ```
> 
> ```
> private ReviewActivity getReviewActivity() {
>     return (ReviewActivity) requireActivity();
> }
> ```

I want to mock it because the getReviewActivity in turn calls getMedia. So I want to mock the media object returned by getMedia call.

**Comment by madhurgupta10** [comment]
It is better to use Whitebox and replace the media object with a mock for ReviewActivity
```
public Media getMedia() {
    return media;
}
```

**Comment by devarsh-mavani-19** [comment]
@nicolas-raoul I have increased code coverage can you please review this PR? Thanks

**Comment by devarsh-mavani-19** [comment]
@4D17Y4 can you please review this? Thanks

**Comment by nicolas-raoul** [comment]
Thanks Devarsh!

Strange that the codecov bot is not showing its check on the latest commits... any idea Madhur maybe?

**Comment by nicolas-raoul** [comment]
I think I found the report: https://codecov.io/github/commons-app/apps-android-commons/commit/6f0b68bb6c9d2dc501407d43187c80ff4b1d3dd1

It seems like the `Finds non-hidden categories from Media instance` block could use better coverage if possible?

**Comment by madhurgupta10** [comment]
Not sure what you are trying to test here but actual subscriptions don't take place in the simulated test environment.

You can make a new method for this and test it separately.

```
 // Finds non-hidden categories from Media instance
                    for(String key : media.getCategoriesHiddenStatus().keySet()) {
                        Boolean value = media.getCategoriesHiddenStatus().get(key);
                        // If non-hidden category is found then set hasNonHiddenCategories to true
                        // so that category review cannot be skipped
                        if(!value) {
                            hasNonHiddenCategories = true;
                            break;
                        }
                    }
```

**Comment by devarsh-mavani-19** [comment]
@nicolas-raoul I have added more coverage can u please check it. The username is different as I was logged with different account locally with different git username, so sorry about that. 😅

File: app/src/main/res/values-de/strings.xml [hunk 10]
```diff
@@ -667,4 +667,5 @@
   <string name="no_location_found_message">Wie wäre es, den Ort hinzuzufügen, an dem das Bild aufgenommen wurde?\nOrtsangaben helfen Wiki-Benutzern, dein Bild zu finden, und machen es so viel nützlicher.\nDanke schön!</string>
   <string name="add_location">Standort hinzufügen</string>
   <string name="feedback_sharing_data_alert">Bitte entferne aus dieser E-Mail alle Informationen, die du nicht öffentlich mitteilen möchtest. Bitte beachte auch, dass deine E-Mail-Adresse, mit der du schreibst, sowie der dazugehörige Name und das Profilbild öffentlich sichtbar sind.</string>
+  <string name="copyright_popup">Bitte lade nur Bilder hoch, die du selbst aufgenommen hast. Wer urheberrechtlich geschützte Bilder hochlädt, wird gesperrt. Dies gilt auch für die Beta-Variante. Vielen Dank, dass du die App getestet hast!</string>
```
Comment (nicolas-raoul) [inline app/src/main/res/values-de/strings.xml:670]: I guess these changes are unrelated?
Would you mind rebasing from master?
If nothing else is working, you can also create a new branch from master, apply the right changes, and send a new pull request :-)

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]

File: app/src/main/res/values-de/strings.xml (hunk 10 above)
Comment (devarsh-mavani-19) [reply app/src/main/res/values-de/strings.xml:670]: @nicolas-raoul those changes are from some else's PR. I actually by mistake changed the author that's why it is showing as if i have made those changes. But actually I merged the master branch and changed the author probably that's why it is showing in my commit.

Comment (nicolas-raoul) [reply app/src/main/res/values-de/strings.xml:670]: Would you mind creating from scratch a new branch and a new pull request?
